cmd/youtubedr/main.go

```diff
@@ -0,0 +1,141 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"flag"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	ytdl "github.com/kkdai/youtube/downloader"
+	"github.com/olekukonko/tablewriter"
+	"golang.org/x/net/proxy"
+)
+
+const usageString string = `Usage: youtubedr [OPTION] [URL]
+Download a video from youtube.
+Example: youtubedr -o "Campaign Diary".mp4 https://www.youtube.com/watch\?v\=XbNghLqsVwU
+`
+
+var (
+	outputFile         string
+	outputDir          string
+	outputQuality      string
+	socks5Proxy        string
+	itag               int
+	info               bool
+	insecureSkipVerify bool
+)
+
+func main() {
+	if err := run(); err != nil {
+		fmt.Fprintf(os.Stderr, "%+v", err)
+		os.Exit(1)
+	}
+}
+
+func run() error {
+	flag.Usage = func() {
+		fmt.Println(usageString)
+		flag.PrintDefaults()
+	}
+	usr, _ := user.Current()
+	flag.StringVar(&outputFile, "o", "", "The output file")
+	flag.StringVar(&outputDir, "d",
+		filepath.Join(usr.HomeDir, "Movies", "youtubedr"),
+		"The output directory.")
+	flag.StringVar(&outputQuality, "q", "", "The output file quality (hd720, medium)")
+	flag.StringVar(&socks5Proxy, "p", "", "The Socks 5 proxy, e.g. 10.10.10.10:7878")
+	flag.IntVar(&itag, "i", 0, "Specify itag number, e.g. 13, 17")
+	flag.BoolVar(&info, "info", false, "show info of video")
+	flag.BoolVar(&insecureSkipVerify, "insecure-skip-tls-verify", false, "skip server certificate verification")
+
+	flag.Parse()
+
+	if len(flag.Args()) == 0 {
+		flag.PrintDefaults()
+		return nil
+	}
+
+	httpTransport := &http.Transport{
+		IdleConnTimeout:       60 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+	}
+
+	if socks5Proxy != "" {
+		log.Println("Using SOCKS5 proxy", socks5Proxy)
+		dialer, err := proxy.SOCKS5("tcp", socks5Proxy, nil, proxy.Direct)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "can't connect to the proxy:", err)
+			os.Exit(1)
+		}
+
+		// set our socks5 as the dialer
+		dc := dialer.(interface {
+			DialContext(ctx context.Context, network, addr string) (net.Conn, error)
+		})
+		httpTransport.DialContext = dc.DialContext
+	} else {
+		httpTransport.DialContext = (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext
+	}
+
+	if insecureSkipVerify {
+		log.Println("Skip server certificate verification")
+		httpTransport.TLSClientConfig = &tls.Config{
+			InsecureSkipVerify: true,
+		}
+	}
+
+	dl := ytdl.Downloader{
+		OutputDir: outputDir,
+	}
+	dl.Client.HTTPClient = &http.Client{Transport: httpTransport}
```

Comment:
It seems to me that exposing Client struct may not be necessary.
Could we just use client as embedded struct  to let Downloader is the only one exported Struct
and provide constructor of  Downloader to initialize client and outputdir?

Reply:
BTW, this comment is  the last one  I would like to discuss, then v2 could be merged and released

Reply:
Oh yes, that's a good idea.